Add MutationRecord Properties (#22072)
mdn/content: files/en-us/web/api/mutationrecord/type/index.md

{{APIRef("DOM")}}

The {{domxref("MutationRecord")}} read-only property **`type`** is the type of the {{domxref("MutationRecord")}} observed by a {{domxref("MutationObserver")}}.

## Value

The property is set to the type of the mutation as a string. The value can be one of the following:

- `attributes` if the mutation was an attribute mutation.

- `characterData` if it was a mutation to a {{domxref("CharacterData")}} node.

- `childList` if the mutation a mutation to the tree of nodes.

## Examples

### Log the type of a mutation

The following example gives you two buttons to manipulate the DOM. The first button adds a new node to the example, and the second button changes the `color` attribute of all the added nodes. A {{domxref("MutationObserver")}} is created to observe it all, and the observer is set to log the `type` of the mutation record to `#log`.

You'll notice the when you add a node, the `type` is `childList`, and when you change the `color` attribute, the `type` is `attributes`.

#### HTML

```html
<button id="add-nodes">Add a node</button>
<button id="set-attribute">Change the color</button>

<button id="reset">Reset</button>

<pre id= "log">Mutation type:</pre>

> <div id= "target" class="my-class"></div>

Reviewer: ```suggestion
<div id= "target"></div>
```

(sorry, my fault!)

```

```css hidden
#log {
  border: 1px dotted black;
  padding: .5rem;
}

.blue {
  color: blue;
}

.red {
 color: red;
}
```

#### JavaScript

```js
const addNodes = document.querySelector("#add-nodes");
const setAttribute = document.querySelector("#set-attribute");
const reset = document.querySelector("#reset");
const log = document.querySelector("#log");
const target = document.querySelector("#target");
let nodeNumber = 1;

addNodes.addEventListener("click", () => {
  const newPara = document.createElement("p");
  newPara.textContent = `Node number: ${nodeNumber}`;
  nodeNumber++;
  target.appendChild(newPara);
});

setAttribute.addEventListener("click", () => {
  if (target.getAttribute("class") === "red") {
    target.setAttribute("class", "blue");
  } else {
    target.setAttribute("class", "red");
  }
});

reset.addEventListener("click", () => self.location.reload());

function logNextSibling(records) {
  for (const record of records) {
    log.textContent = `Mutation type: ${record.type}`;
  }
}

> const observer = new MutationObserver(logNextSibling);

Reviewer: ```suggestion
const observer = new MutationObserver(logMutationType);
```

> function logNextSibling(records) {
>   for (const record of records) {
>     log.textContent = `Mutation type: ${record.type}`;
>   }
> }
> 
> const observer = new MutationObserver(logNextSibling);

Reviewer: ```suggestion
function logMutationType(records) {
  for (const record of records) {
    log.textContent = `Mutation type: ${record.type}`;
  }
}

const observer = new MutationObserver(logMutationType);
```

---
title: MutationRecord.type
slug: Web/API/MutationRecord/type
page-type: web-api-instance-property
tags:
  - type
  - MutationRecord
  - Property
  - Reference
browser-compat: api.MutationRecord.type
---

{{APIRef("DOM")}}

> The {{domxref("MutationRecord")}} property **`type`** is the type of the {{domxref("MutationRecord")}} observed by a {{domxref("MutationObserver")}}.

Reviewer: ```suggestion
The {{domxref("MutationRecord")}} read-only property **`type`** is the type of the {{domxref("MutationRecord")}} observed by a {{domxref("MutationObserver")}}.
```

## Value

> The property is set to the type of the mutated node as a string. The value can be one of the following:

Reviewer: ```suggestion
The property is set to the type of the mutation as a string. The value can be one of the following:
```

> - The value will be `attributes` if the mutation to the target node of the was an attribute mutation, {{domxref("Attr")}}. This mutated node type can appear if the "attribute" parameter of {{domxref("MutationObserver.observe()")}} is set to true or the `attributeFilter` parameter is in use.
> 
> - The value will be `childList` if the mutation to the target node was a mutation to the tree of nodes, i.e. the `childList` paremeter of {{domxref("MutationObserver.observe()")}} is set to true.
> 
> - The value will be `characterData` if it was called on a mutation to a {{domxref("CharacterData")}} node.

Reviewer: ```suggestion
- `attributes` if the mutation was an attribute mutation.

- `characterData` if it was a mutation to a {{domxref("CharacterData")}} node.

- `childList` if the mutation a mutation to the tree of nodes.
```

## Examples

> This code snippet is used to observe changes in a userlist: the userlist is DOM element that functions as a list of all users currently connected to a server (in this example, consider it is a `<ul>`). The userlist itself contains child elements (i.e. `<li>`), each containing an attribute associated with the username the user status.

Reviewer: Consider a live sample based on this pen: https://codepen.io/willbamberg/pen/wvXPoPB ?
